libcnb/src/runtime.rs

```diff
@@ -122,17 +126,32 @@ fn cnb_runtime_build<B: Buildpack>(buildpack: &B) -> Result<(), B::Error> {
     let buildpack_plan =
         read_toml_file(&args.buildpack_plan_path).map_err(Error::CannotReadBuildpackPlan)?;
 
-    let context = BuildContext {
-        layers_dir,
+    let build_outcome = buildpack.build(BuildContext {
+        layers_dir: layers_dir.clone(),
         app_dir,
         stack_id,
         platform,
         buildpack_plan,
         buildpack_dir: read_buildpack_dir()?,
         buildpack_descriptor: read_buildpack_toml()?,
-    };
+    })?;
+
+    match build_outcome.0 {
+        InnerBuildOutcome::Fail => process::exit(1),
```

Comment:
Yes, this is now. There was no way of failing the build other than `exit` or returning an `Err` from `build`. I added the possibility just to avoid surprises for the user. If `BuildOutcomeBuilder` has a `success` method, why doesn't it have a `fail`? Especially since `DetectOutcomeBuilder` has one.

I don't think buildpack authors would want to use this often. The main advantage to using `Ok(BuildOutcomeBuilder::fail().build())` instead of exiting yourself would be the abstraction of the error code. Even though unlikely, the expected exit code for a failed build in the CNB spec might change. If you exit manually, you'd need to adjust. By letting libcnb handle this, buildpack authors don't need to care what the correct exit code is or if it changed between spec versions.

If we're talking semantics, returning an `Err` from `build` is just like letting an exception bubble up when it can't be meaningfully handled anyway. Failing the build explicitly is different in that regard. I can imagine (corner) cases where you'd want to fail the build not because of an error but another condition. It's a blurry line though.

WRT error/message: I'd say this is up to the author to handle this case.

Reply:
I guess the reason I ask is that I had thought `buildpack.handle_error()` would be where the exit code / handling would occur (after the error has propagated back up to `cnb_runtime()`, but it's sounding like the buildpack author should also be catching all errors inside the buildpack too?

In which case we now have two places performing error handling? (Buildpacks still need to think about the top level `handle_error` due to libcnb-specific errors that can occur).

Reply:
@edmorley and I discussed this on a call. At this point, there is not much value in failed `BuildOutcome`s. We currently model build failures with `Result` and adding a second way to deal with failed builds might be confusing or at least requires additional boilerplate to ensure consistent handling of failures.

Especially since we could not come up with great use-cases for failing a build via `BuildOutcome`, we decided it's best to remove it for now.